File: Javascript/FrogJump.js
/**
 * Link: https://leetcode.com/problems/frog-jump/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[]} stones
 * @return {boolean}
 */
var canCross = function(stones) {
    var map = {};
    var trace = {};
    
    for (var i = 0; i < stones.length; i++) {
        map[stones[i]] = i;
    }
    
    for (pos of stones) {
        trace[pos] = [];    
    }
    
    return dfs(stones, 0, 0, map, trace);
    //return dfs(stones, 0, 0);
};


function dfs(stones, start, jump, map, trace) {
    // if the jum of current position is wrong then return false
    if (trace[stones[start]].indexOf(jump) > -1) return false; 
    if (start == stones.length - 1) return true;
    
    var canJump = false;
    var nextJumps = jump == 0 ? 
                [jump + 1] : [jump - 1, jump, jump + 1];
    
    for (var nextJump of nextJumps) {
        var nextStart = stones[start] + nextJump;
        var iNextStart = map[nextStart];
        
        if (iNextStart != undefined) {
            canJump = dfs(stones, iNextStart, nextJump, map, trace);
            if (canJump) return true;
            
            // save the wrong nextJump for each pos
            trace[nextStart].push(nextJump);
        }
    }
    
    return canJump;
}

File: C#/FrogJump.cs
/*
 * Problem: https://leetcode.com/problems/frog-jump/submissions/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public bool CanCross(int[] stones) {
        var traces = new Dictionary<int, List<int>>();
        
        foreach(var stone in stones) {
            traces.Add(stone, new List<int>());
        }
        
        var canJump = Backtrack(stones, 0, 0, traces);
        
        return canJump;
    }
    
    public bool Backtrack(int[] stones, int current, int jump, Dictionary<int, List<int>> traces) {
        if (current == stones[stones.Length - 1]) 
            return true;
        
        var nextJumps = new int[] { jump - 1, jump, jump + 1 };
        foreach(var nextJump in nextJumps) {
            if (nextJump > 0 
                && nextJump + jump > 0 // the case both of them is zero 
                && stones.Any(stone => stone == current + nextJump) // can land on the next stone
                && current + nextJump <= stones[stones.Length - 1] // dont pass the last stone
                && !traces[current].Any(nJump => nJump == nextJump) // dont have the wrong jump
                && Backtrack(stones, current + nextJump, nextJump, traces))
                return true;
            
            // save the wrong pair stone - jump
            traces[current].Add(nextJump);
        }
        
        return false;
    }
}